File: MomentumiOS/MomentumiOS/Presentation/Meals/MealViewModel.swift
//
//  MealViewModel.swift
//  MomentumiOS
//
//  Created by Mwai Banda on 8/14/23.
//  Copyright © 2023 Momentum. All rights reserved.
//

import Foundation
import TinyDi
import MomentumSDK

class MealViewModel: ObservableObject {
    @Inject private var mealController: MealController
    @Inject private var notificationController: NotificationController

    func getMeals(onCompletion: @escaping ([Meal]) -> Void) {
        mealController.getAllMeals { res in
            if let meals = res.data as? [Meal] {
                onCompletion(meals)
            }
        }
    }
    
    func postMeal(request: MealRequest, onCompletion: @escaping (Meal) -> Void) {
        mealController.postMeal(request: request) { [weak self] res in
            if let meal = res.data {
                onCompletion(meal)
                self?.notificationController.sendNotification(notification: Notification(
                    title: "Momentum Church: Indiana",
                    body: "We have a new meal, organised by \(meal.user.fullname) for the \(meal.recipient)",
                    topic: MultiplatformConstants.shared.ALL_USERS_TOPIC
                )) { res in
                    if let data = res.data {
                        print("Notification posted successfully")
                    }
                }
            }
        }
    }
    func postVolunteeredMeal(request: VolunteeredMealRequest, onCompletion: @escaping (VolunteeredMeal) -> Void) {
        mealController.postVolunteeredMeal(request: request) { res in
            if let meal = res.data {
                onCompletion(meal)
            }
        }
    }

}

File: MomentumAndroid/src/main/java/com/mwaibanda/momentum/android/presentation/meals/MealViewModel.kt
package com.mwaibanda.momentum.android.presentation.meals

import android.util.Log
import androidx.lifecycle.ViewModel
import com.mwaibanda.momentum.data.mealDTO.MealRequest
import com.mwaibanda.momentum.data.mealDTO.VolunteeredMealRequest
import com.mwaibanda.momentum.domain.controller.MealController
import com.mwaibanda.momentum.domain.controller.NotificationController
import com.mwaibanda.momentum.domain.models.Meal
import com.mwaibanda.momentum.domain.models.Notification
import com.mwaibanda.momentum.domain.models.VolunteeredMeal
import com.mwaibanda.momentum.utils.MultiplatformConstants
import com.mwaibanda.momentum.utils.Result

class MealViewModel(
    private val mealController: MealController,
    private val notificationController: NotificationController
): ViewModel() {
    fun getMeals(onCompletion: (List<Meal>) -> Unit) {
        mealController.getAllMeals {
            when (it) {
                is Result.Failure -> {
                    Log.e("MealViewModel[getAllMeals]", it.message ?: "" )
                }
                is Result.Success -> {
                    onCompletion(it.data ?: emptyList())
                }
            }
        }
    }

    fun postMeal(request: MealRequest, onCompletion: (Meal) -> Unit) {
        mealController.postMeal(request) {
            when (it) {
                is Result.Failure -> {
                    Log.e("MealViewModel[postMeal]", it.message ?: "" )
                }
                is Result.Success -> {
                    it.data?.let { meal ->
                        onCompletion(meal)
                        notificationController.sendNotification(
                            notification = Notification(
                                title = "Momentum Church: Indiana",
                                body = "We have a new meal, organised by ${meal.user.fullname} for the ${meal.recipient}",
                                topic = MultiplatformConstants.ALL_USERS_TOPIC
                            )
                        ){}
                    }

                }
            }
        }
    }

    fun postVolunteeredMeal(request: VolunteeredMealRequest, onCompletion: (VolunteeredMeal) -> Unit) {
        mealController.postVolunteeredMeal(request) {
            when (it) {
                is Result.Failure -> {
                    Log.e("MealViewModel[postVolunteeredMeal]", it.message ?: "" )
                }
                is Result.Success -> {
                    it.data?.let(onCompletion)
                }
            }
        }
    }
}